raft.go

```diff
@@ -284,7 +284,13 @@ func (r *Raft) runCandidate() {
 	metrics.IncrCounter([]string{"raft", "state", "candidate"}, 1)
 
 	// Start vote for us, and set a timeout
-	voteCh := r.electSelf()
+	var voteCh <-chan *voteResult
+	var prevoteCh <-chan *voteResult
+	if r.preVote {
+		prevoteCh = r.electSelf(true)
+	} else {
+		voteCh = r.electSelf(false)
+	}
```

Comment:
At first I felt like we need to somehow keep the state around so that if we timeout a PreVote campaign we fall back to a real one next time... but then I realise that would defeat the purpose of a prevote because a partitioned server would only prevote for one round before it resumed incrementing it's term...

I do wonder a bit about whether it's possible to get "stuck" never quite winning a pre vote in time but also not making progress.

Here's the scenario I came up with.

 * Three servers A, B, C
 * Only A is upgraded to have PreVote
 * B is currently leader
 * C is down for a bit
 * A disconnects temporarily
 * B looses leadership and becomes a follower waiting for an election (or to start one)
 * A starts election first and use PreVote
 * B treats it like RealVote and so grants it an resets its contact time (which prevents it from timing out and becoming a candidate
 * A sees the request granted by B but it's not enough for a quorum (using the code here) so keeps waiting
 * A's prevote campaign times out and it starts again... electionTimeout is always less than or equal to heartbeat timeout so this could continue forever even though there are a quorum of healthy servers.

Note that this scenario is "fixed" if we use the suggestion I made below about only requiring `quorum - 1` real vote grants and then upgrading our own. But it's an example of something non-obvious that could cause the cluster to get stuck in a whole new way that isn't possible today which makes me 🤔.

Reply:
What about B in that case, at some point B will start an election too and because B don't have prevote A will vote for B and we will be out of the loop. 
The idea, is that:
- if we have majority of nodes with prevote, pre vote will be guaranteed to go through
- if we have majority of nodes without prevote, the prevote will become a vote
- if we have same number, the prevote will continue until: a) we have another node that support prevote added, b) a node that don't support prevote take over
I think that the quorum - 1 idea would work but it will be an optimization.

Reply:
You just solved a bug I was chasing with one of the tests, when I set the number of nodes to 2 prevotes,1 no prevote 😄

Reply:
While going through this use case to fix it I think it's part of a more general use case. Where `grantedVotes + grantedPreVotes >= quorum` but `grantedVotes < quorum` and `grantedPreVotes < quorum`.
I think the best way to solve this would be to transition to election as soon as `grantedVotes + grantedPreVotes >= quorum`.